README.md
# Wrapper: Offline
After learning that GoAnimate Wrapper (a project by VisualPlugin) would be shut down, Wrapper: Offline was created. (simply called "Offline" for the rest of this page) Unlike the original project, Offline can not be shut down or "patched". You have the files all on your computer. This project is important for historical and archival purposes, as the ability to use this legacy editor and themes would be completely gone without it. Besides simply emulating the original Flash editor, some additional features and improvements have been made. And Offline will continue to get additional updates in the future.

## Running / Installation
To start Offline on Windows, open start_wrapper.bat. It'll automate just about everything for you and, well, start Wrapper: Offline. On your first run, you will likely need to right-click it and click "Run as Administrator". This allows it to properly install what it needs to run. After your initial run, you shouldn't need to do that again, you can start it as normal.

If you want to import videos and characters from the original Wrapper or any other clones of it, open its folder and drag the "_SAVED" folder into Offline's "wrapper" folder. If you have already made any videos or characters, this will not work. Please only import on a new install with no saved characters or videos, or take the "_SAVED" folder in Offline out before dragging the old one in. If you want to import character IDs from the original LVM, you can paste `&original_asset_id=[ID HERE]` at the end of the link for the matching character creator.

## Updates & Support
For support, the first thing you should do is read through faq.txt, it most likely has what you want to know. If you can't find what you need, you can join the [Discord server](https://discord.gg/Kf7BzSw). Joining the server is recommended, as there is a whole community to help you out. If you don't use Discord, you can email mailbenson@protonmail.com to get in contact with me directly, but don't expect nearly as quick of a response.

## Dependencies
This program relies on Flash, Node.js and http-server to work properly. SilentCMD is also used to suppress all the extra logging noise you'd only need for troubleshooting and development. These all have been included with the project (utilities folder) to ensure full offline operation and will be installed if missing. The "wrapper" folder and http-server have their own dependencies, but they are included as well.

## License
Most of this project is free/libre software[1] under the MIT license. You have the freedom to run, change, and share this as much as you want.
This includes:
  - Files in the "wrapper" folder
  - Batch files made for Wrapper: Offline
  - Node.js
  - http-server
  - SilentCMD
  - Chromium Web Store

ungoogled-chromium is under the BSD 3-Clause license, which grants similar rights, but has some differences from MIT. MediaInfo has a similar BSD 2-Clause license. 7zip's license is mostly LGPL, but some parts are under the BSD 3-clause License, and some parts have an unRAR restriction. Stylus is under the GNU GPLv3 license. These licenses can be found in each program's folder in utilities\sourcecode.

The source code for compiled programs are all stored in utilities\sourcecode, and you can modify these as you wish. Parts of Offline that run from their source code directly (such as batch scripts) are not included in that folder, for obvious reasons.

Flash Player (utilities folder) and GoAnimate's original assets (server folder) are proprietary and do not grant you these rights, but if they did, this project wouldn't need to exist. Requestly, an addon included in Offline's browser, is sadly proprietary software, but you're free to remove the Chromium profile and use a fresh one if this bothers you. Requestly is primarily included because of how popular it is with our community.

While completely unnecessary, if you decide to use your freedom to change the software, it would be greatly appreciated if you sent it to me so I can implement it into the main program! With credit down here of course :)

## Credits
**Please do not contact anyone on the list for support, use the Discord server.**

Original Wrapper credits:
| Name         | Contribution         |
| ------------ | -------------------- |
| VisualPlugin | GoAnimate Wrapper    |
| Imageny      | Custom/Modded Assets |
| CLarramore   | Bug fixes            |
No members of the original team are officially working on Offline, even if they have contributed. Some members of the original team have asked to not be given credit, and they have been removed.

Developers will have (D) on their name, to show they are officially working on Wrapper: Offline. Everyone else has simply contributed something at some point.
Wrapper: Offline credits:
| Name            | Contribution                     |
| --------------- | -------------------------------- |
| Benson (D)      | Wrapper: Offline                 |
| NathanTDA       | Logo                             |
| Blukas/GoTube   | Import idea, Mega Comedy World 2 |
| VisualPlugin    | Text to speech voices            |
| Imageny         | Text to speech voices            |
| creepyjokes2000 | Waveform fix, improved actions   |
| JoshAnimate     | Headgear fix                     |

These are unaffiliated people that they haven't directly done anything for the project (and probably don't even know it exists) but still deserve credit for their things. Kinda like a shoutout but in a project's readme. ***Please do not contact them about Wrapper: Offline.***
| Name               | Contribution                     |
| ------------------ | -------------------------------- |
| Vyond              | Creators of the themes we love   |
| http-party         | Creators of http-server          |
| Stephan Brenner    | Creator of SilentCMD             |
| vocatus            | Some of TronScript's batch code  |
| ss64.com           | Incredible CMD info resource     |
| robvanderwoude.com | Also amazing CMD info resource   |

## Footnotes
[1] - See <https://www.gnu.org/philosophy/free-sw.html> for a better definition of free software.


changelog.md
<!-- markdownlint-disable MD004 MD024 -->
# Wrapper: Offline Changelog
This changelog is based off [Keep a Changelog](https://keepachangelog.com) and ~~somewhat but not really~~ adheres to [Semantic Versioning](https://semver.org/). If you have no idea what that means, basically this changelog is formatted in a specific consistent way, and version numbers are formatted as X.Y.Z, X being a major not-backwards-compatible update, Y being a feature update, and X being bug fixes.

## Version 1.2.3 - 2021-03-06 - i am HUNGRY FOR UPDATES!!!
### Added
  - Character browser
### Changed
  - Centered character creator
### Fixed
  - Remove the Voicery TTS engine following their shutdown
  - Voiceforge fix
  - fullscreen in video player

## Version 1.2.2 - 2020-07-06 - sufferingggggg
### Changed
  - CD commands are now PUSHD commands
  - Open Node.js/http-server scripts show helpful messages when closing
### Fixed
  - Import script now detects whether the file directory has quotes
  - Import script no longer loops when asking questions
  - Import script specifically only checks files in the base folder, no subfolders
  - Upgrade comes with node_modules folder to stop MODULE_NOT_FOUND errors
  - Taskkill errors are now muted

## Version 1.2.1 - 2020-07-06 - weeeeee bug fixes
### Changed
  - Upgrade script just checks for file existence instead of using variables
### Fixed
  - Importer now copies to _THEMES folder
  - Reverted use of CHOICE command because people were having issues with it
  - Punch action now doesn't look horrible (credit to creepyjokes2000)
  - Launcher creates checks folder if nonexistent
  - Removed chcp 65001 from files that don't need it to prevent crashing
  - HTTPS cert now checks for the right file
  - Old music/sfx are back

## Version 1.2.0 - 2020-07-05 - a launcher overhaul disguised as the file importing update
### What do I care about
This update basically has file uploading. New TTS voices have been added, including Eric and Jennifer. The headgear section works now. Waveforms work now. Mega Comedy World 2 has been added. An FAQ is now included with Wrapper. The launcher is far more stable now, as you can tell from all the entries that start with the word Launcher.
### Added
  - File uploading!!!...kinda. It's more of a workaround.
  - Loads of new TTS voices (yes this means eric and jennifer)
  - Mega Comedy World 2 by Blukas/GoTube
  - Functioning waveforms (credit to creepyjokes2000)
  - Headgear sections (credit to JoshAnimate)
  - Frequently Asked Questions file/page (#answers from the discord server)
  - Disclaimer when launching for the first time
  - Folder in utilities that stores files to remember certain events
  - Added script for resetting your install (meant for developers)
  - Dry run option for testing launcher easier
  - Non-admin HTTPS cert install option (but admin is still preferred)
  - Source code of programs is now included in utilities folder
  - Chromium now includes Requestly and Stylus extensions
  - Copy of JPEXS Free Flash Decompiler included
  - Upgrade script
### Changed
  - Turned on "use gpu" in Flash attributes. May or may not do anything, who knows
  - Settings is now interactive (text version still in utilities\config.bat, but you probably shouldn't mess with it)
  - Heavily modified the README (mostly simplifying and moving things to faq.txt)
  - Video list mostly redone
  - Server page redone
  - HTTPS cert now also applies to server ports both 4664 and 4343
  - HTTPS cert now expires on 9999-12-31 instead of 4343-04-03, just to be extra sure that by the time it expires technology will have advanced so far that HTTPS will only exist as a relic of history and most if not all of Wrapper: Offline's functions will only be possible in an emulated version of today's technology. if that prediction turns out to be completely false someone can just renew it and be done so 10,000s kids can still make grounded videos on youtube
  - Node.js installer is now 12.18.1
  - Flash installer is now 32.0.0.371
  - Launcher explicitly mentions verbose mode is on while idle
  - Launcher checks for HTTPS cert check more reliably
  - Launcher variables now use y/n instead of yes/no
  - Launcher admin required prompt specifies what is needed
  - Launcher single key choices now use choice command
  - Launcher open README command now opens FAQ
  - Launcher idle options reorganized
  - Launcher server page option now opens in Chromium if enabled
  - Launcher turns on skipping dependency check if all are found
### Fixed
  - Sprint action looks less broken (credit to creepyjokes2000)
  - patch. it's in the utilities folder, don't move it next to the launcher
  - HTML injection was possible via the Upload a Movie button
  - More voices should now fit within the character limit
  - All scripts use UTF-8 to prevent special character crashes
  - Server's port is now 4664 instead of 4343, Wrapper port is 4343 instead of 80 (hopefully will prevent EADDRINUSE errors)
  - Launcher now closes any running Node.js processes (hopefully also prevents EADDRINUSE)
  - Launcher should no longer crash from running as Admin
  - Launcher now checks if it's in a Wrapper: Offline folder
  - Launcher waits before opening your browser to allow Wrapper to start up
  - Launcher now kills Edge and Brave when installing Flash
  - Launcher can now generate settings if not available
  - Launcher now finds more Flash versions and is more aware of settings
  - Launcher now checks if installers exist before running them
  - Launcher stays open when crashing to be troubleshooted
  - Launcher no longer crashes from spaces in answers
  - Launcher now switches drives when moving to W:O folder
  - Launcher variables now use delayed expansion syntax

wrapper/server.js
const env = Object.assign(process.env,
	require('./env'),
	require('./config'));

const http = require('http');
const chr = require("./character/redirect");
const pmc = require('./character/premade');
const chl = require('./character/load');
const chs = require('./character/save');
const mvu = require('./movie/upload');
const asu = require('./asset/upload');
const stl = require('./static/load');
const stp = require('./static/page');
const asl = require('./asset/load');
const asL = require('./asset/list');
const ast = require('./asset/thmb');
const mvl = require('./movie/load');
const mvL = require('./movie/list');
const mvm = require('./movie/meta');
const mvs = require('./movie/save');
const mvt = require('./movie/thmb');
const thL = require('./theme/list');
const thl = require('./theme/load');
const tsv = require('./tts/voices');
const tsl = require('./tts/load');
const evt = require('./events');
const url = require('url');

const functions = [
	mvL,
	pmc,
	asl,
	evt,
	chl,
	thl,
	thL,
	chs,
	asL,
	tsl,
	chr,
	ast,
	mvm,
	mvl,
	mvs,
	mvt,
	tsv,
	asu,
	mvu,
	stp,
	stl,
];

module.exports = http.createServer((req, res) => {
	const parsedUrl = url.parse(req.url, true);
	//if (!parsedUrl.path.endsWith('/')) parsedUrl.path += '/';
	const found = functions.find(f => f(req, res, parsedUrl));
	if (!found) { res.statusCode = 404; res.end(); }
}).listen(env.PORT || env.SERVER_PORT, console.log);
// 2epik4u is weird
// update: still weird

wrapper/tts/main.js
const voices = require('./info').voices;
const get = require('../request/get');
const qs = require('querystring');
const https = require('https');

module.exports = function (voiceName, text) {
	return new Promise((res, rej) => {
		const voice = voices[voiceName];
		switch (voice.source) {
			case 'polly': {
				var buffers = [];
				var req = https.request({
					hostname: 'pollyvoices.com',
					port: '443',
					path: '/api/sound/add',
					method: 'POST',
					headers: {
						'Content-Type': 'application/x-www-form-urlencoded',
					},
				}, r => {
					r.on('data', b => buffers.push(b));
					r.on('end', () => {
						var json = JSON.parse(Buffer.concat(buffers));
						if (json.file)
							get(`https://pollyvoices.com${json.file}`).then(res);
						else
							rej();
					});
				});
				req.write(qs.encode({ text: text, voice: voice.arg }));
				req.end();
				break;
			}
			case 'cepstral':
			case "cepstral": {
                https.get('https://www.cepstral.com/en/demos', r => {
                    const cookie = r.headers['set-cookie'];
                    var q = qs.encode({
                        voice: voice.arg,
                        voiceText: text,
                        rate: 170,
                        pitch: 1,
                        sfx: 'none',
                    });
                    var buffers = [];
                    var req = https.get({
                        host: 'www.cepstral.com',
                        path: `/demos/createAudio.php?${q}`,
                        headers: { Cookie: cookie },
                        method: 'GET',
                    }, r => {
                        r.on('data', b => buffers.push(b));
                        r.on('end', () => {
                            var json = JSON.parse(Buffer.concat(buffers));
                            get(`https://www.cepstral.com${json.mp3_loc}`).then(res).catch(rej);
                        })
                    });
                });
                break;
            }
			case 'vocalware': {
				var q = qs.encode({
					EID: voice.arg[0],
					LID: voice.arg[1],
					VID: voice.arg[2],
					TXT: text,
					IS_UTF8: 1,
					HTTP_ERR: 1,
					ACC: 3314795,
					API: 2292376,
					vwApiVersion: 2,
					CB: 'vw_mc.vwCallback',
				});
				var req = https.get({
					host: 'cache-a.oddcast.com',
					path: `/tts/gen.php?${q}`,
					method: 'GET',
					headers: {
						Referer: 'https://www.vocalware.com/index/demo',
						Origin: 'https://www.vocalware.com',
						'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36',
					},
				}, r => {
					var buffers = [];
					r.on('data', d => buffers.push(d));
					r.on('end', () => res(Buffer.concat(buffers)));
					r.on('error', rej);
				});
				break;
			}
			case 'watson': {
				var q = qs.encode({
					text: text,
					voice: voice.arg,
					download: true,
					accept: "audio/mp3",
				});
				console.log(https.get({
					host: 'text-to-speech-demo.ng.bluemix.net',
					path: `/api/v1/synthesize?${q}`,
					headers: {
						Referer: 'https://www.vocalware.com/index/demo',
						Origin: 'https://www.vocalware.com',
						'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36',
					},
				}, r => {
					var buffers = [];
					r.on('data', d => buffers.push(d));
					r.on('end', () => res(Buffer.concat(buffers)));
					r.on('error', rej);
				}));
				break;
			}
		}
	});
}


wrapper/tts/load.js
const voices = require("./info").voices;
const qs = require("querystring");
const brotli = require("brotli");
const https = require("https");
const md5 = require("js-md5");
const base64 = require("js-base64");
const http = require("http");

// Fallback option for compatibility between Wrapper and https://github.com/Windows81/Text2Speech-Haxxor-JS.
let get;
try {
	get = require("../misc/get");
} catch {
	get = require("./get");
}

module.exports = (voiceName, text) => {
	return new Promise(async (res, rej) => {
		const voice = voices[voiceName];
		switch (voice.source) {
			case "polly1": {
                var req = https.request({
                        hostname: "voicemaker.in",
						port: "443",
                        path: "/voice/standard",
                        method: "POST",
                        headers: {
							"content-type": "application/json",
							cookie: "__cfduid=dea5a80a27d2fbaa14bb2f3c006e2b6401612476329; connect.sid=s%3AaPvH6lwiFtj6xbRGpiXUtRObjiNSJdUE.0IoDTJibP7XtLMTLt3DGFlOfv4Nee%2BN%2B0eQvsr9fWm0; __stripe_mid=0a37ccc8-cc13-474d-97da-aa0a5f9f47398bdf7a; __stripe_sid=d7b84afb-50aa-4813-8e33-d69585f09dd94ce8c7",
							"csrf-token": "",
                            origin: "https://voicemaker.in",
							referer: "https://voicemaker.in/",
							"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.101 Safari/537.36",
                            "x-requested-with": "XMLHttpRequest",
                        },
                    },
                    (r) => {
						var buffers = [];
						r.on("data", (b) => buffers.push(b));
                        r.on("end", () => {
							var json = JSON.parse(Buffer.concat(buffers));
							get(`https://voicemaker.in${json.path}`).then(res).catch(rej);
						});
						r.on("error", rej);
					});
					req.write(`{"Engine":"standard","Provider":"ai101","OutputFormat":"mp3","VoiceId":"${voice.arg}","LanguageCode":"${voice.language}-${voice.country}","SampleRate":"22050","effect":"default","master_VC":"advanced","speed":"0","master_volume":"0","pitch":"0","Text":"${text}","TextType":"text","fileName":""}`);
					req.end();
					break;
			}
			case "pollyNeural": {
                var req = https.request({
                        hostname: "voicemaker.in",
						port: "443",
                        path: "/voice/standard",
                        method: "POST",
                        headers: {
							"content-type": "application/json",
							cookie: "__cfduid=dea5a80a27d2fbaa14bb2f3c006e2b6401612476329; connect.sid=s%3AaPvH6lwiFtj6xbRGpiXUtRObjiNSJdUE.0IoDTJibP7XtLMTLt3DGFlOfv4Nee%2BN%2B0eQvsr9fWm0; __stripe_mid=0a37ccc8-cc13-474d-97da-aa0a5f9f47398bdf7a; __stripe_sid=d7b84afb-50aa-4813-8e33-d69585f09dd94ce8c7",
							"csrf-token": "",
                            origin: "https://voicemaker.in",
							referer: "https://voicemaker.in/",
							"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.101 Safari/537.36",
                            "x-requested-with": "XMLHttpRequest",
                        },
                    },
                    (r) => {
						var buffers = [];
						r.on("data", (b) => buffers.push(b));
                        r.on("end", () => {
							var json = JSON.parse(Buffer.concat(buffers));
							get(`https://voicemaker.in${json.path}`).then(res).catch(rej);
						});
						r.on("error", rej);
					});
					req.write(`{"Engine":"neural","Provider":"ai101","OutputFormat":"mp3","VoiceId":"${voice.arg}","LanguageCode":"${voice.language}-${voice.country}","SampleRate":"22050","effect":"default","master_VC":"advanced","speed":"0","master_volume":"0","pitch":"0","Text":"${text}","TextType":"text","fileName":""}`);
					req.end();
					break;
			}
            case "polly": {
                var buffers = [];
                var req = https.request({
                        hostname: "pollyvoices.com",
                        port: "443",
                        path: "/api/sound/add",
                        method: "POST",
                        headers: {
                            "Content-Type": "application/x-www-form-urlencoded",
                        },
                    },
                    (r) => {
                        r.on("data", (b) => buffers.push(b));
                        r.on("end", () => {
                            var json = JSON.parse(Buffer.concat(buffers));
                            if (json.file) get(`https://pollyvoices.com${json.file}`).then(res);
                            else rej();
                        });
                    }
                );
                req.write(qs.encode({
                    text: text,
                    voice: voice.arg
                }));
                req.end();
                break;
            }
			case "cepstral": {
				https.get('https://www.cepstral.com/en/demos', r => {
					const cookie = r.headers['set-cookie'];
					var q = qs.encode({
						voiceText: text,
						voice: voice.arg,
						createTime: 666,
						rate: 170,
						pitch: 1,
						sfx: 'none',
					});
					var buffers = [];
					var req = https.get({
						host: 'www.cepstral.com',
						path: `/demos/createAudio.php?${q}`,
						headers: { Cookie: cookie },
						method: 'GET',
					}, r => {
						r.on('data', b => buffers.push(b));
						r.on('end', () => {
							var json = JSON.parse(Buffer.concat(buffers));
							get(`https://www.cepstral.com${json.mp3_loc}`).then(res).catch(rej);
						})
					});
				});
				break;
			}
			case "vocalware": {
				var [eid, lid, vid] = voice.arg;
				var cs = md5(`${eid}${lid}${vid}${text}1mp35883747uetivb9tb8108wfj`);
				var q = qs.encode({
					EID: voice.arg[0],
					LID: voice.arg[1],
					VID: voice.arg[2],
					TXT: text,
					EXT: "mp3",
					IS_UTF8: 1,
					ACC: 5883747,
					cache_flag: 3,
					CS: cs,
				});
				var req = https.get(
					{
						host: "cache-a.oddcast.com",
						path: `/tts/gen.php?${q}`,
						headers: {
							Referer: "https://www.oddcast.com/",
							Origin: "https://www.oddcast.com/",
							"User-Agent":
								"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36",
						},
					},
					(r) => {
						var buffers = [];
						r.on("data", (d) => buffers.push(d));
						r.on("end", () => res(Buffer.concat(buffers)));
						r.on("error", rej);
					}
				);
				break;
			}
			case "watson": {
				var q = qs.encode({
					text: text,
					voice: voice.arg,
					download: true,
					accept: "audio/mp3",
				});
				https.get(
					{
						host: "text-to-speech-demo.ng.bluemix.net",
						path: `/api/v3/synthesize?${q}`,
					},
					(r) => {
						var buffers = [];
						r.on("data", (d) => buffers.push(d));
						r.on("end", () => res(Buffer.concat(buffers)));
						r.on("error", rej);
					}
				);
				break;
			}
			/* case "acapela": {
				var buffers = [];
				var acapelaArray = [];
				for (var c = 0; c < 15; c++) acapelaArray.push(~~(65 + Math.random() * 26));
				var email = `${String.fromCharCode.apply(null, acapelaArray)}@gmail.com`;
				var req = https.request(
					{
						hostname: "acapelavoices.acapela-group.com",
						path: "/index/getnonce",
						method: "POST",
						headers: {
							"Content-Type": "application/x-www-form-urlencoded",
						},
					},
					(r) => {
						r.on("data", (b) => buffers.push(b));
						r.on("end", () => {
							var nonce = JSON.parse(Buffer.concat(buffers)).nonce;
							var req = http.request(
								{
									hostname: "acapela-group.com",
									port: "8080",
									path: "/webservices/1-34-01-Mobility/Synthesizer",
									method: "POST",
									headers: {
										"Content-Type": "application/x-www-form-urlencoded",
									},
								},
								(r) => {
									var buffers = [];
									r.on("data", (d) => buffers.push(d));
									r.on("end", () => {
										const html = Buffer.concat(buffers);
										const beg = html.indexOf("&snd_url=") + 9;
										const end = html.indexOf("&", beg);
										const sub = html.subarray(beg, end).toString();
										http.get(sub, (r) => {
											r.on("data", (d) => buffers.push(d));
											r.on("end", () => {
												res(Buffer.concat(buffers));
											});
										});
									});
									r.on("error", rej);
								}
							);
							req.end(
								qs.encode({
									req_voice: voice.arg,
									cl_pwd: "",
									cl_vers: "1-30",
									req_echo: "ON",
									cl_login: "AcapelaGroup",
									req_comment: `{"nonce":"${nonce}","user":"${email}"}`,
									req_text: text,
									cl_env: "ACAPELA_VOICES",
									prot_vers: 2,
									cl_app: "AcapelaGroup_WebDemo_Android",
								})
							);
						});
					}
				);
				req.end(
					qs.encode({
						json: `{"googleid":"${email}"`,
					})
				);
				break;
			} */
            case "acapela": {
                var buffers = [];
                var req = https.request({
                        hostname: "acapela-box.com",
                        path: "/AcaBox/dovaas.php",
                        method: "POST",
                        headers: {
                            "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
							Cookie: "AcaBoxLogged=logged; AcaBoxUsername=goaniwrap; acabox=92s39r5vu676g5ekqehbu2o0f2; AcaBoxFirstname=Keegan",
							Origin: "https://acapela-box.com",
                            Referer: "https://acapela-box.com/AcaBox/index.php",
							"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.101 Safari/537.36",
                        },
                    },
                    (r) => {
                        r.on("data", (b) => buffers.push(b));
                        r.on("end", () => {
                            var json = JSON.parse(Buffer.concat(buffers));
							get(`${json.snd_url}`).then(res).catch(rej);
                        });
                    }
                );
                req.write(qs.encode({
                    text: text,
                    voice: voice.arg,
					listen: 1,
					format: "MP3",
					codecMP3: 1,
					spd: 180,
					vct: 100,
					byline: 0,
					ts: 666
                }));
                req.end();
                break;
            }
			case "acapelaOld": {
				var q = qs.encode({
					inputText: base64.encode(text),
				});
				https.get(
					{
						host: "voice.reverso.net",
						path: `/RestPronunciation.svc/v1/output=json/GetVoiceStream/voiceName=${voice.arg}?${q}`,
					},
					(r) => {
						var buffers = [];
						r.on("data", (d) => buffers.push(d));
						r.on("end", () => res(Buffer.concat(buffers)));
						r.on("error", rej);
					}
				);
				break;
			}
			case "svox": {
				var q = qs.encode({
					apikey: "e3a4477c01b482ea5acc6ed03b1f419f",
					action: "convert",
					format: "mp3",
					voice: voice.arg,
					speed: 0,
					text: text,
					version: "0.2.99",
				});
				https.get(
					{
						host: "api.ispeech.org",
						path: `/api/rest?${q}`,
					},
					(r) => {
						var buffers = [];
						r.on("data", (d) => buffers.push(d));
						r.on("end", () => res(Buffer.concat(buffers)));
						r.on("error", rej);
					}
				);
				break;
			}
			case "festival": {
                var req = https.request({
                        hostname: "texttomp3.online",
                        path: "/php/logic/textToSpeech.php",
                        method: "POST",
                        headers: {
                            "Content-Type": "application/x-www-form-urlencoded; charset=UTF8",
							Host: "www.texttomp3.online",
							Origin: "https://www.texttomp3.online",
							Referer: "https://www.texttomp3.online/",
							"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.101 Safari/537.36",
                        },
                    },
                    (r) => {
						var buffers = [];
						r.on("data", (b) => buffers.push(b));
                        r.on("end", () => {
							const mp3 = Buffer.concat(buffers).toString();
							const loc = `https://www.texttomp3.online/audio_tmp/${mp3}`
							get(loc).then(res).catch(rej);
						});
						r.on("error", rej);
					});
                    req.write(qs.encode({
                        msg: text,
                        voice: voice.arg,
						usebackmusic: 0,
						backmusicfile: "",
						backmusicvolume: ""
                    }));
					req.end();
					break;
			}
			case "wavenet": {
                var req = https.request({
                        hostname: "texttospeechapi.wideo.co",
                        path: "/api/wideo-text-to-speech",
                        method: "POST",
                        headers: {
                            "Content-Type": "application/json",
							Origin: "https://texttospeech.wideo.co",
							Referer: "https://texttospeech.wideo.co/",
							"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.101 Safari/537.36",
                        },
                    },
                    (r) => {
						var buffers = [];
						r.on("data", (b) => buffers.push(b));
                        r.on("end", () => {
							var json = JSON.parse(Buffer.concat(buffers));
							get(`${json.url}`).then(res).catch(rej);
						});
						r.on("error", rej);
					});
					req.write(`{"data":{"text":"${text}","speed":1,"voice":"${voice.arg}"}}`);
					req.end();
					break;
			}
			/*
			case "acapela": {
				var q = qs.encode({
					cl_login: "VAAS_MKT",
					req_snd_type: "",
					req_voice: voice.arg,
					cl_app: "seriousbusiness",
					req_text: text,
					cl_pwd: "M5Awq9xu",
				});
				http.get(
					{
						host: "vaassl3.acapela-group.com",
						path: `/Services/AcapelaTV/Synthesizer?${q}`,
					},
					(r) => {
						var buffers = [];
						r.on("data", (d) => buffers.push(d));
						r.on("end", () => {
							const html = Buffer.concat(buffers);
							const beg = html.indexOf("&snd_url=") + 9;
							const end = html.indexOf("&", beg);
							const sub = html.subarray(beg + 4, end).toString();
							if (!sub.startsWith("://")) return rej();
							get(`https${sub}`).then(res).catch(rej);
						});
						r.on("error", rej);
					}
				);
				break;
			}
			*/
			case "readloud": {
				const req = https.request(
					{
						host: "readloud.net",
						path: voice.arg,
						method: "POST",
						port: "443",
						headers: {
							"Content-Type": "application/x-www-form-urlencoded",
						},
					},
					(r) => {
						var buffers = [];
						r.on("data", (d) => buffers.push(d));
						r.on("end", () => {
							const html = Buffer.concat(buffers);
							const beg = html.indexOf("/tmp/");
							const end = html.indexOf(".mp3", beg) + 4;
							const sub = html.subarray(beg, end).toString();
							const loc = `https://readloud.net${sub}`;
							get(loc).then(res).catch(rej);
						});
						r.on("error", rej);
					}
				);
				req.end(
					qs.encode({
						but1: text,
						but: "Submit",
					})
				);
				break;
			}
			case "cereproc": {
				const req = https.request(
					{
						hostname: "www.cereproc.com",
						path: "/themes/benchpress/livedemo.php",
						method: "POST",
						headers: {
							"content-type": "text/xml",
							"accept-encoding": "gzip, deflate, br",
							origin: "https://www.cereproc.com",
							referer: "https://www.cereproc.com/en/products/voices",
							"x-requested-with": "XMLHttpRequest",
							cookie: "Drupal.visitor.liveDemo=666",
						},
					},
					(r) => {
						var buffers = [];
						r.on("data", (d) => buffers.push(d));
						r.on("end", () => {
							const xml = String.fromCharCode.apply(null, brotli.decompress(Buffer.concat(buffers)));
							const beg = xml.indexOf("https://cerevoice.s3.amazonaws.com/");
							const end = xml.indexOf(".mp3", beg) + 4;
							const loc = xml.substr(beg, end - beg).toString();
							get(loc).then(res).catch(rej);
						});
						r.on("error", rej);
					}
				);
				req.end(
					`<speakExtended key='666'><voice>${voice.arg}</voice><text>${text}</text><audioFormat>mp3</audioFormat></speakExtended>`
				);
				break;
			}
			/*
			case "sestek": {
				https.get(
					{
						hostname: "ttsdemo.sestek.com",
						path: "/Demo.aspx",
					},
					(r) => {
						var cookie = r.headers["set-cookie"];
						https.get(
							{
								hostname: "ttsdemo.sestek.com",
								headers: { Cookie: cookie },
								path: "/Demo.aspx",
							},
							(r) => {
								var buffers = [];
								r.on("data", (d) => buffers.push(d));
								r.on("end", () => {
									html = Buffer.concat(buffers).toString();
									var vs = /__VIEWSTATE" value="([^"]+)/.exec(html);
									var vg = /__VIEWSTATEGENERATOR" value="([^"]+)/.exec(html);
									var ev = /__EVENTVALIDATION" value="([^"]+)/.exec(html);

									if (vs && ev && vg) {
										vs = vs[1];
										vg = vg[1];
										ev = ev[1];
									} else rej();

									var q = qs.encode({
										__EVENTTARGET: "Button1",
										__EVENTARGUMENT: "",
										__VIEWSTATE: vs,
										__EVENTVALIDATION: ev,
										__VIEWSTATEGENERATOR: vg,
										ddlVoices: voice.arg,
										TextBox1: text,
									});

									const req = https.request(
										{
											hostname: "ttsdemo.sestek.com",
											path: "/demo.aspx",
											method: "POST",
											headers: {
												cookie: cookie,
												"Content-Type": "application/x-www-form-urlencoded",
												"Content-Length": q.length,
											},
										},
										(r) => {
											var buffers = [];
											r.on("data", (d) => buffers.push(d));
											r.on("end", () => {
												console.log(r.headers.location);
												get(r.headers.location).then(res).catch(rej);
											});
											r.on("error", rej);
										}
									);
									req.end(q);
								});
								r.on("error", rej);
							}
						);
					}
				);
				break;
			}
			*/
		}
	});
};
